src/include/denc.h

```diff
@@ -307,6 +349,14 @@ inline void denc_signed_varint(T& v, bufferptr::iterator& p)
 inline void denc_varint_lowz(uint64_t v, size_t& p) {
   p += sizeof(v) + 2;
 }
+
+inline void denc_varint_lowz_exact(uint64_t v, size_t& p) {
+  int lowznib = v ? (ctz(v) / 4) : 0;
+  if (lowznib > 3)
+    lowznib = 3;
+  v >>= lowznib * 4 - 2;
```

Comment:
I'm afraid we aren't on the same page. I thought about implementing the function with `ctz` and `cbits` supplemented by some *plain* math (additions, subtractions, ands, ors) -- **without** divisions, multiplications nor even (variable, `CL`-related) shifts.

I disagree about the costs of shifts in comparison to the mentioned *plain* math instructions. According to the [Intel® 64 and IA-32 Architectures Optimization Reference Manual](http://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-optimization-manual.html) variable shifts on newer Intel's microarchitectures are translated into 2 or 3 microops:

> **3.5.1.6  Variable Bit Count Rotation and Shift**
>
> In Intel microarchitecture code name Sandy Bridge, The “`ROL/ROR/SHL/SHR reg, cl`” instruction has  three micro-ops. When the flag result is not needed, one of these micro-ops may be discarded, providing better performance in many common usages.

AFAIK the basic (register-to-register) variants of `SUB`, `ADD`, `AND`, `OR` are reflected in one microop only. Moreover, nowadays CPUs tend to have less resources for shifts in general. For instance, Haswell and and later microarchitectures have only two ports and execution units able to deal with shifts while four are avaible for the initially mentioned microops [1].

These restrictions affect the CPU throughput. Variable shift costs around 1.5 cycles, non-variable (immediate) takes 0.5 while a *plain* math instruction is 0.25 only [2].

Anyway, I'm not sure whether we need to focus on such low-level optimizations rgiht now as there are much more painful bottlenecks to address first.

[1]. Table 2-7.  "Haswell Microarchitecture Execution Units and Representative Instructions", page 40
[2].  Table C-17. "General Purpose Instructions", pages 663 and 664